### PR TITLE
Default torch DataLoader num_workers to 1

### LINKED ISSUE #1560: PyTorch Distributed Training: File descriptors opened and never closed
When using distributed training w/ PyTorch v2.30, RETURNN opens file descriptors every subepoch that are then never closed.

Comparing the output of `lsof -l` for the same process w/ a few subepochs difference one gets diffs like this:
image

with blocks of FDs like below repeating:

```
pt_main_t 13339 17144239  100r     FIFO               0,13      0t0             2311492 pipe
pt_main_t 13339 17144239  101w     FIFO               0,13      0t0             2311492 pipe
pt_main_t 13339 17144239  102u  a_inode               0,14        0               13849 [eventpoll]
pt_main_t 13339 17144239  103u     IPv4            2311493      0t0                 TCP g-02.apptek.local:33695 (LISTEN)
pt_main_t 13339 17144239  104u     IPv4            2311497      0t0                 TCP g-02.apptek.local:33695->localhost:34500 (ESTABLISHED)
```

Using the pipe FD 101 in the example above and some quick and dirty tracing code (see below) that I wrapped around various functions in the Pytorch engine I was able to narrow down the pipe FD's construction to

```
-> Engine.train_epoch
-> DataLoader.__iter__
-> DataLoader._get_iterator
-> _SingleProcessDataLoaderIter.__init__
-> _BaseDataLoaderIter.__init__
-> torch.distributed.new_group(backend="gloo")
-> torch.distributed._new_group_with_tag
-> torch.distributed._new_process_group_helper
-> backend_class = ProcessGroupGloo(backend_prefix_store, group_rank, group_size, timeout=timeout)
-> native code
```

The only reason for existence of this process group is for distributing a random seed between the worker group members. Note, in this example it does not make a difference if we use `num_workers > 0` or not as both the `_SingleProcessDataLoaderIter` and `_MultiProcessDataLoaderIter` inherit from `_BaseDataLoaderIter`. What I'm wondering about is why it's not destructed and the FDs closed once the data loader iter goes out of scope.

Torch allows re-using workers using the "persistent_workers"-flag, and in that case the workers are not recreated every subepoch but just reset, which means Torch won't create new process groups and the FDs won't accumulate. This can be a workaround for now.

lsof-Diff with `num_workers=1, persistent_workers=True` (notice the missing green blocks of added lines visible in the other diff): 
image

EDIT: I found out we already default to setting `persistent_workers=True` when `num_workers` is set. So in the case with `torch_dataloader_opts = {"num_workers": 1}` we are good.

----

Tracing code:

```python
from .basic import is_valid_fd
from returnn.log import log

num_fd = 101


def fd(name):

    def decorator(func):
        name_ = name or func.__name__

        def unfold(val):
            for v in val:
                yield v
                print(f"after value {name_}, FD {num_fd} is {is_valid_fd(num_fd)}", file=log.v3)

        def wrapped(*args, **kwargs):
            print(f"before {name_}, FD {num_fd} is {is_valid_fd(num_fd)}", file=log.v3)
            try:
                val = func(*args, **kwargs)
                return unfold(val) if hasattr(val, "__next__") else val
            finally:
                print(f"after {name_}, FD {num_fd} is {is_valid_fd(num_fd)}", file=log.v3)

        return wrapped

    return decorator


from contextlib import contextmanager


@contextmanager
def fd_c(name):
    try:
        print(f"before {name}, FD {num_fd} is {is_valid_fd(num_fd)}", file=log.v3)
        yield
    finally:
        print(f"after {name}, FD {num_fd} is {is_valid_fd(num_fd)}", file=log.v3)
```

### PR DESCRIPTION
**Pull request by NeoLegends**
Closes #1560 

This is a pretty small, but significant change, so I did make this into a PR and did not directly commit to master.

### CODE REVIEW

**Review by albertz — APPROVED** [review]

**Comment by albertz** [comment]
Ok, two failing tests. Can you check what's the problem there?

**Review by albertz — CHANGES_REQUESTED** [review]
Check/fix tests.

**Comment by NeoLegends** [comment]
Sure, asap

**Comment by albertz** [comment]
I meant `test_torch_engine_train` and `test_torch_engine_train_exception`. I think sth with serialization fails, probably due to the change in this PR. So this is important. (The question is also whether potentially some real existing setups will break now for similar reasons.)

**Comment by NeoLegends** [comment]
So, it seems like setups that have `nn.Module` classes declared in a non-global scope will break. I don't think many people have setups like that though, as we normally use imports to declare `get_model`.

**Comment by NeoLegends** [comment]
Can't (un)pickle local classes or functions.

**Comment by albertz** [comment]
> including the config's `get_model`, which in the test case points to a locally (i.e. in the scope of the test function) declared class.

Ah that's how it got the reference, I see. I wonder though if this might be problematic at some point. Because for the dataset, we don't care about `get_model` or many other things from the config. (Actually, why do we care at all about the config?)
